### PR TITLE
Python 3 compatibility

### PR DESCRIPTION
**Pull request by svenstaro**
We used six for compatibility and removed all the decoding/encoding stuff as well as tests for it. It made compatibility a whole lot harder first and foremost but it also doesn't add anything to the modern web. If we assume unicode with utf-8 (an assumption that Python 3 makes anyway) then we don't even need the different encodings. All in all, we think it's reasonable to assume that we feed unicode to this module at all times instead of dealing with it on a one-to-one basis.

### CODE REVIEW

**Comment by svenstaro** [comment]
This closes #1 on merge.

**Comment by euangoddard** [comment]
This mainly looks good, but I'm concerned about some of the removal of code around encoding. This was one of the challenges we faced around ensuring that the data transmitted from the remote host could be decoded.

I'm not aware of exactly how `six` works, but I don't think some of the changes will address the decoding of the response. Moreover, I don't understand why tests have been removed. Surely these can be made to run under python3?

**Comment by byteslicer** [comment]
I can understand the problem, but shouldn't the programmer handle the input encoding conversion?

**Comment by euangoddard** [comment]
@lazzrek that is a good point. The guy who wrote most of this library is on vacation at the moment. Let me chat to him as soon as he's back and hopefully we can resolve this issue.

File: recaptcha.py [hunk 1]
```diff
@@ -109,132 +92,131 @@ def __init__(
...
         This method communicates with the remote reCAPTCHA API and uses the
         ``verification_timeout`` set in the constructor.
-        
+
         """
         if not solution_text or not challenge_id:
             return False
-        
-        solution_text_decoded = \
-            solution_text.decode(RECAPTCHA_CHARACTER_ENCODING)
```
Comment (gnarea) [inline recaptcha.py:186]: This cannot be removed because the resulting value is passed to urlencode(), which should only receive bytestrings, not unicode strings which could contain non-ASCII characters.

You suggest in a comment that we let users of this library deal with it, but:
- Developers using this library would have to convert their users' input to a UTF-8 decoded bytestring, each time they call this method. In my opinion, that's unnecessary hassle that we can easily avoid.
- That would be a backwards incompatible change. And unless we absolutely have to do it, I'd rather avoid such changes because they'll just annoy developers. Everybody using this library would have to update their code.

The call to six.u() doesn't really have any effect. The input already is unicode.

File: recaptcha.py [hunk 2]
```diff
@@ -247,31 +229,30 @@ def _get_recaptcha_response_for_solution(
             })
         request = Request(
             url=verification_url,
-            data=request_data,
+            data=request_data.encode('utf-8'),
```
Comment (gnarea) [inline recaptcha.py:232]: request_data is output from urlencode(), which can only return return ASCII strings, so I think this in unnecessary because the only contentious value would be the "solution_text", which should be decoded anyway as mentioned above.

File: recaptcha.py [hunk 3]
```diff
@@ -247,31 +229,30 @@ def _get_recaptcha_response_for_solution(
...
         if self.verification_timeout is not None:
             urlopen_kwargs['timeout'] = self.verification_timeout
         try:
             response = urlopen(request, **urlopen_kwargs)
-        except URLError, exc:
+        except URLError as exc:
             raise RecaptchaUnreachableError(exc)
         else:
-            response_lines = response.read().splitlines()
+            response_lines = response.read().decode('utf-8').splitlines()
```
Comment (gnarea) [inline recaptcha.py:244]: Why did you decode from UTF-8?

I don't think we have to worry about encoding because this is just an ASCII string that says "true" or "false", but if anything, we should be _encoding_ to UTF-8.

File: recaptcha.py [hunk 4]
```diff
@@ -247,31 +229,30 @@ def _get_recaptcha_response_for_solution(
...
         is_solution_correct = response_lines[0] == 'true'
         verification_result = {'is_solution_correct': is_solution_correct}
         if not is_solution_correct:
             verification_result['error_code'] = response_lines[1]
-        
-        return verification_result
 
+        return verification_result
 
-#{ Exceptions
```
Comment (gnarea) [inline recaptcha.py:273]: This grouping is part of our conventions, so we'll revert this change when we apply your pull request.

File: recaptcha.py [hunk 5]
```diff
@@ -90,16 +73,16 @@
...
     """Stateless reCAPTCHA client."""
-    
+
     def __init__(
         self,
         private_key,
         public_key,
         recaptcha_options=None,
         verification_timeout=None,
-        ):
```
Comment (gnarea) [inline recaptcha.py:100]: This indentation style is part of our coding conventions, so we'll revert this change when we apply your pull request

File: requirements.txt [hunk 6]
```diff
@@ -0,0 +1 @@
+six
```
Comment (gnarea) [inline requirements.txt]: This file is redundant because this dependency is already specified in setup.py

File: setup.py [hunk 7]
```diff
@@ -20,7 +20,7 @@
 
 _branch_path = os.path.abspath(os.path.dirname(__file__))
 _readme = open(os.path.join(_branch_path, 'README.txt')).read()
-_version = open(os.path.join(_branch_path, 'VERSION.txt')).readline().rstrip()
```
Comment (gnarea) [inline setup.py:23]: I don't think this change was necessary to make the library compatible with Python 3, and should be reverted because our Sphinx documentation relies on the VERSION.txt file that you removed.

**Comment by gnarea** [comment]
Hi there, Sven!

Thank you so much for contributing to the project! We're all bound to be using Python 3 at some point, so it's indeed nice to see some of our own libraries compatible with it.

I think the use of "six" makes a lot of sense, but perhaps most of the changes were unnecessary to make the library compatible with Python 3? Many changes reverted our coding conventions and the test suite was practically rewritten... I'm not sure all of that was necessary.

GitHub reports that 7 files were changed with 263 additions and 365 deletions, but I'm having a hard time working out what didn't work in Python 3. Can you please explain what exactly didn't work?

Thanks!

File: recaptcha.py (hunk 3 above)
Comment (byteslicer) [inline recaptcha.py:244]: in python3 response.read() returns a byte object not a string

File: recaptcha.py (hunk 5 above)
Comment (byteslicer) [inline recaptcha.py:100]: In some cases that indentation style is not pep8 conform

File: recaptcha.py (hunk 3 above)
Comment (gnarea) [inline recaptcha.py:244]: That's right... And in Python 2, that method returns the counterpart: An "str" object.

But I still don't know why you decoded the value. As I said above, it's all ASCII, and if anything we should be encoding to UTF-8 instead of decoding from UTF-8.

splitlines() can be called on Python 2 str objects and Python 3 byte strings.

File: recaptcha.py (hunk 5 above)
Comment (svenstaro) [inline recaptcha.py:100]: If you guys really want to disregard pep8, then we will revert those changes if it results in a merge.

File: recaptcha.py (hunk 4 above)
Comment (svenstaro) [inline recaptcha.py:273]: You sure you wouldn't rather have a pep8 conform public package? I think it makes it easier for people to get started on it.

**Comment by svenstaro** [comment]
The tests were rewritten to be more efficient using mock. This way we can shave of lots of code that would otherwise have had to write ourselves.

File: recaptcha.py (hunk 5 above)
Comment (gnarea) [inline recaptcha.py:100]: Can you please refer me to the point in PEP-8 that we're not complying with?

I just read the sections about "Maximum Line Length" and "Indentation", and I couldn't see anything that says that this should not be done. So, to the best of my knowledge, PEP-8 has no "opinion" about this convention of ours.

File: recaptcha.py (hunk 4 above)
Comment (gnarea) [inline recaptcha.py:273]: Of course we want to comply with PEP-8, and we do. 

To the best of my knowledge, this point is fine as far as PEP-8 is concerned. Unless you can refer me to a point that says it isn't?

We have to bear in mind that every developer and every development team will have conventions that build on top of PEP-8. And that doesn't mean that they're not complying with PEP-8. It's impossible for a document like PEP-8 to cover 100% of all code styles, so additional conventions are needed to fill the blanks.

**Comment by gnarea** [comment]
Thanks for the clarification, Sven! That makes sense, but to follow up on that: What was that code that you would've had to write? Wouldn't the pre-existing test suite work just fine (perhaps with some minor tweaks)?

But more importantly, what exactly isn't working in Python 3?

File: recaptcha.py (hunk 5 above)
Comment (byteslicer) [inline recaptcha.py:100]: I can not find anything either in the pep8 docs but the pep8 checker marks it as wrong: http://pep8online.com/s/1sTQEERF

Another way to fix it is to indent the parameters 2 times:
http://pep8online.com/s/OIAbngK2

**Comment by svenstaro** [comment]
There we some syntax errors as well as problems with bytes vs strings.

We ended up writing our own 10 lines of request code to handle validation of a recaptcha as we didn't really need the whole scope of this project and trying to get our changes merged ended up taking longer than expected and we didn't want to maintain our own fork for a long term.

So, I'm sorry to say but this concludes our efforts on this pull request. We will still leave it up and give you a chance to merge it or parts of it if you like. Our general recommendations are: Use the pep8 utility, use mock, use unittest (instead of nose), use six and you should be able to simplify your code and unify it across Python versions.

**Comment by svenstaro** [comment]
Sooo... are you going to merge this?

**Comment by gnarea** [comment]
Hello, Sven.

We'd be very pleased to merge a pull request that is limited to adding support for Python 3. Unfortunately, your pull request changes way too many things that are irrelevant and which we don't agree with (e.g., introduces backwards incompatible changes, removes tests that we consider important, undoes some of our coding conventions).

I really appreciate the time you've put into making this library better, but I'm afraid we cannot merge it in its current state for the reasons above. If you're willing to create a new pull request that is strictly limited to adding Python 3 support, I'd be more than happy to look at it.

Cheers.

**Comment by svenstaro** [comment]
I do see where your disagreement comes from and agree that this had better been split into separate commits. However, we seem to agree on one fundamental change: The decoding/encoding stuff. We think that for all practical reasons, the programmers shouldn't worry about that, especially since python2/python3 handle this differently.

Internally, we have now done the recaptcha handling with a small and maintainable function that I'm going to attach here for reference. Therefore, we're not interested in pursuing the maintenance of this library any further. Though I, for one, am a bit sad to see it rot like this.

I think it should be modernized by ripping out many of the legacy things to make it more maintainable and approachable.

The function we use:

https://gist.github.com/04b7c0ec8421ff8b2c64
